### PR TITLE
add agent output parsers

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

File: libs/langchain/langchain/agents/output_parsers/openai_functions.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    If one is not passed, then the AIMessage is assumed to be the final output.
+    """
+
+    @property
+    def _type(self) -> str:
+        return "openai-functions-agent"
+
+    @staticmethod
+    def _parse_ai_message(message: BaseMessage) -> Union[AgentAction, AgentFinish]:
```
Comment (nfcampos) [inline libs/langchain/langchain/agents/output_parsers/openai_functions.py:41]: what's the advantage v being a standalone function?

Comment (hwchase17) [reply libs/langchain/langchain/agents/output_parsers/openai_functions.py:41]: could be, main thing was just making sure to use it here and the openai functions agent. could make public function not on class. no preference really

File: libs/langchain/langchain/agents/output_parsers/json.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    ```
+    {
+      "action": "Final Answer",
+      "action_input": "4"
+    }
+    ```
+    """
+
+    pattern = re.compile(r"```(?:json)?\n(.*?)```", re.DOTALL)
```
Comment (baskaryan) [inline libs/langchain/langchain/agents/output_parsers/json.py:42]: should there be a \n before finishing ``` as well?

File: libs/langchain/langchain/agents/output_parsers/openai_functions.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        """Parse an AI message."""
+        if not isinstance(message, AIMessage):
+            raise TypeError(f"Expected an AI message got {type(message)}")
+
+        function_call = message.additional_kwargs.get("function_call", {})
+
+        if function_call:
+            function_name = function_call["name"]
+            try:
+                _tool_input = json.loads(function_call["arguments"])
```
Comment (baskaryan) [inline libs/langchain/langchain/agents/output_parsers/openai_functions.py:46]: should we strict=False here as well?

File: libs/langchain/langchain/agents/output_parsers/react_json_single_input.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    being returned.
+
+    ```
+    Thought: agent thought here
+    Final Answer: The temperature is 100 degrees
+    ```
+
+    """
+
+    pattern = re.compile(r"^.*?`{3}(?:json)?\n(.*?)`{3}.*?$", re.DOTALL)
```
Comment (baskaryan) [inline libs/langchain/langchain/agents/output_parsers/react_json_single_input.py:43]: can we just have one pattern constant somewhere that all the json markdown things use

**Review by baskaryan — APPROVED** [review]

**Comment by hwchase17** [comment]
good comments @baskaryan - but this is just porting existing logic so want to that in this pr. if we want to improve also fine with that but will do in follow ups

**Pull request merged**